### PR TITLE
Flashattn and multiround chat

### PR DESCRIPTION
**Pull request by Louym**
New optimization for context stage. Update the readme in TinyChat.

### CODE REVIEW

**Comment by casper-hansen** [comment]
Dear @Louym, I invite you to upstream these optimizations into AutoAWQ as well. We have a largely reworked attention module that could use some of the changes from your great work in this PR:

https://github.com/casper-hansen/AutoAWQ/blob/main/awq/modules/fused/attn.py

**Comment by Louym** [comment]
@casper-hansen Thanks for your invitation. I will if I have time.

File: tinychat/models/llama.py [hunk 1]
```diff
@@ -259,9 +265,10 @@ def forward(
         start_pos: int,
         freqs_cis: torch.Tensor,
         mask: Optional[torch.Tensor],
+        chunk_prefill:bool,
     ):
         h = x + self.self_attn.forward(
-            self.input_layernorm(x), start_pos, freqs_cis, mask
+            self.input_layernorm(x), start_pos, freqs_cis, mask,chunk_prefill
```
Comment (ys-2020) [inline tinychat/models/llama.py:271]: Please leave a blank space between comma and `chunk_prefill`. Same for the rest of the modifications.

File: tinychat/modules/fused_attn.py [hunk 2]
```diff
@@ -303,9 +312,153 @@ def forward(
             output = output.reshape(bsz, 1, -1)
 
         return self.o_proj(output)
+    
+class QuantLlamaAttentionFused_flashattn_singlebatch(nn.Module):
```
Comment (ys-2020) [inline tinychat/modules/fused_attn.py:316]: Please do not use `_` in the class name.

File: tinychat/modules/fused_vision_attn.py [hunk 3]
```diff
@@ -222,16 +222,92 @@ def forward(
...
+        attn_output = self.out_proj(attn_output)
+
+        return attn_output, attn_weights
+    
 def make_fused_vision_attn(model, dev):
     """
     Replace all LlamaAttention modules with QuantLlamaAttention modules, fusing the q, k, v projections.
     """
     model = model.cpu()
+    print(model)
```
Comment (ys-2020) [inline tinychat/modules/fused_vision_attn.py:306]: Do we need to print model here?

File: tinychat/modules/fused_vision_attn.py [hunk 4]
```diff
@@ -256,6 +332,11 @@ def make_fused_vision_attn(model, dev):
             attn = CLIPAttentionFused(
                 m.embed_dim, m.num_heads, qkv_layer, m.out_proj, dev
             )
+        if isinstance(m, SiglipAttention):
+            # attn = CLIPAttentionFused(
+            #     m.embed_dim, m.num_heads, qkv_layer, m.out_proj, dev
+            # )
+            continue
```
Comment (ys-2020) [inline tinychat/modules/fused_vision_attn.py:339]: Hi @Louym. Could you please clarify why we need this?

File: tinychat/utils/prompt_templates.py [hunk 5]
```diff
@@ -80,22 +80,33 @@ def __init__(
...
-            + self.starter
-            + self.role1
-            + ": {prompt}"
-            + self.stopper
-            + self.sen_spliter
-            + self.starter
-            + self.role2
-            + ":"
-        )
+    def update_template(self, outputs,promptcache=0):
```
Comment (ys-2020) [inline tinychat/utils/prompt_templates.py:83]: Do we still have prompt_cache now?

File: tinychat/vlm_demo_new.py [hunk 6]
```diff
@@ -197,20 +211,20 @@ def main(args):
...
+        default="/home/yuming/workspace/llm-awq_1/tinychat/VILA1.5-13b-AWQ/llm/vila-1.5-13b-w4-g128-awq-v2.pt",
     )
     parser.add_argument(
         "--precision", type=str, default="W4A16", help="compute precision"
     )
     parser.add_argument(
         "--image-file",
         type=str,
-        default="https://llava.hliu.cc/file=/nobackup/haotian/code/LLaVA/llava/serve/examples/extreme_ironing.jpg",
+        default="/home/yuming/workspace/VILA/demo_images/av.png",
```
Comment (ys-2020) [inline tinychat/vlm_demo_new.py:227]: Please be careful about the default paths. Ideally the user name should not appear in the paths.

File: tinychat/demo.py [hunk 7]
```diff
@@ -202,8 +215,12 @@ def skip(*args, **kwargs):
     # Optimize AWQ quantized model
     if args.precision == "W4A16" and args.model_type.lower() == "llama":
         from tinychat.modules import make_quant_norm, make_quant_attn
-
-        make_quant_attn(model, args.device)
+        if args.flash_attn:
+            print("Enabling flash-attention!")
+            make_quant_attn(model, args.device,1)
```
Comment (ys-2020) [inline tinychat/demo.py:220]: Please avoid using integer as bool.

File: tinychat/README.md [hunk 8]
```diff
@@ -143,10 +141,28 @@ We recently evaluated AWQ's performance on the Visual Language Models. Here is a
...
 
+## New Optimization of Context Stage
+We have optimized the speed of the context stage and updated our code with several enhancements, including the adoption of flash attention and the elimination of redundant computations. The key optimizations include:
+1. Adopting the flash-attention kernel (currently supports only single-batch operations).
+2. Computing only the last tokens in the final logits layer.
+3. Utilizing history KV caches in the context stage to speed up. (chunk prefill)
+
+Under certain conditions, these improvements can result in up to an 8x speedup in TTFT(Time To the First Token) with minor accuracy loss comparing with the old version of TinyChat. We conducted experiments on the Orin GPU, and the detailed results are provided below. 
+
+![](./figures/TTFT_Speedup_flash_only_last_logits.png)
```
Comment (ys-2020) [inline tinychat/README.md:152]: Could we have a table instead of figures here? Let me know what's your opinion. @kentang-mit

**Comment by ys-2020** [comment]
Hi @Louym,

Thanks for your efforts! The PR generally looks very good to me. I have left some comments in this PR. Could you please have a look?

Additionally, after addressing the issues, perhaps you can use `black` to optimize the code style of the repo (with a separate commit). 

@kentang-mit . Please feel free to add your thoughts.

Thanks,
Shang

File: tinychat/README.md [hunk 9]
```diff
@@ -143,10 +141,28 @@ We recently evaluated AWQ's performance on the Visual Language Models. Here is a
 | FP16      | 84.3  | 64.6 | 62.2   | 87.2  | 73.6  | 87.3 | 1726.82 | 82.4 | 80.2   | 69.1 |
 | AWQ-INT4   | 84.1  | 64.4 | 61.3   | 86.7  | 73.2  | 88.2 | 1714.79 | 83.2 | 79.6   | 68.9 | 
 
+## New Optimization of Context Stage
+We have optimized the speed of the context stage and updated our code with several enhancements, including the adoption of flash attention and the elimination of redundant computations. The key optimizations include:
+1. Adopting the flash-attention kernel (currently supports only single-batch operations).
+2. Computing only the last tokens in the final logits layer.
+3. Utilizing history KV caches in the context stage to speed up. (chunk prefill)
+
+Under certain conditions, these improvements can result in up to an 8x speedup in TTFT(Time To the First Token) with minor accuracy loss comparing with the old version of TinyChat. We conducted experiments on the Orin GPU, and the detailed results are provided below.
```
Comment (ys-2020) [inline tinychat/README.md:150]: Why accuracy loss? I think we are doing exact the same workload as original and should expect no accuracy loss. The numerical error shouldn't be expressed as accuracy loss, which might be misleading.

File: tinychat/README.md [hunk 10]
```diff
@@ -210,6 +226,8 @@ python demo.py --model_type llama \
     --model_path /PATH/TO/LLAMA2/llama-2-7b-chat \
     --precision W16A16
 ```
+You can also try using flash-attention along with chunk prefill. Use the following three arguments when running demo: ```bash
```
Comment (ys-2020) [inline tinychat/README.md:229]: Why three?

File: tinychat/benchmark_context.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    )
+    parser.add_argument(
+        "--max_batch_size", type=int, default=1, help="maximum batch size for kv cache"
+    )
+    parser.add_argument(
+        "--flash_attn",
+        action="store_true",
+        help="whether to use flash attention",
+    )
+    parser.add_argument(#You can only see speedup after the first round!
```
Comment (ys-2020) [inline tinychat/benchmark_context.py:52]: Do you think we should release prompt cache? @kentang-mit

File: tinychat/benchmark_context.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+        "--promptcache",
+        action="store_true",
+        help="Whether to use promptcache. If used, in context stage, the new input questions will not see former, which will lead to speedup but forgetting",
+    )
+    parser.add_argument(
+        "--decodinglike_context",
```
Comment (ys-2020) [inline tinychat/benchmark_context.py:58]: Let's simplify the terminologies. `--decodinglike_context` seems to be confusing.

**Comment by ys-2020** [comment]
I have reviewed the new modifications. It looks good to me. Thanks @Louym

**Comment by ys-2020** [comment]
Thank you @Louym . I will merge this PR.
